File: 203.移除链表元素/203.移除链表元素-java/203.移除链表元素.java
/*
 * @lc app=leetcode.cn id=203 lang=java
 * @lcpr version=30121
 *
 * [203] 移除链表元素
 */

// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode sentinel = new ListNode(205, head);
        sentinel.next = head;
        ListNode cur = sentinel;
        while (cur.next != null) {
            if (cur.next.val == val) {
                cur.next = cur.next.next;
            } else {
                cur = cur.next;
            }
        }
        return sentinel.next;

    }
}
// @lc code=end

/*
 * // @lcpr case=start
 * // [1,2,6,3,4,5,6]\n6\n
 * // @lcpr case=end
 * 
 * // @lcpr case=start
 * // []\n1\n
 * // @lcpr case=end
 * 
 * // @lcpr case=start
 * // [7,7,7,7]\n7\n
 * // @lcpr case=end
 * 
 */

File: 203.移除链表元素/203.移除链表元素-cpp/203.移除链表元素.cpp
/*
 * @lc app=leetcode.cn id=203 lang=cpp
 * @lcpr version=30204
 *
 * [203] 移除链表元素
 */

// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeElements(ListNode *head, int val) // ListNode *是返回类型
    {
        ListNode *sentinelNode = new ListNode(0, head);
        ListNode *cur = sentinelNode;
        while (cur->next != NULL)
        {
            if (cur->next->val == val)
            {
                ListNode *temp = cur->next;
                cur->next = cur->next->next;
                delete temp;
            } else { 
                // 为什么要加else?假设删除的是最后一个节点,那么cur->next就指向null,如果不加else语句继续循环的话,cur就会更新为null,在while判断循环的时候cur->next的时候就会产生错误
                cur = cur->next;
            }
        }
        head = sentinelNode->next;
        delete sentinelNode;
        return head;
    }
};
// @lc code=end

/*
// @lcpr case=start
// [1,2,6,3,4,5,6]\n6\n
// @lcpr case=end

// @lcpr case=start
// []\n1\n
// @lcpr case=end

// @lcpr case=start
// [7,7,7,7]\n7\n
// @lcpr case=end

*/
